client/go/dogma/cmd/util.go

```diff
@@ -0,0 +1,284 @@
+func netrcInfo(machineName string) *netrc.Machine {
+	if usr, err := user.Current(); err == nil {
+		netrcFilePath := usr.HomeDir + string(filepath.Separator)
+		if runtime.GOOS == "windows" {
+			netrcFilePath += "_netrc"
+		} else {
+			netrcFilePath += ".netrc"
+		}
+		if netrc, err := netrc.ParseFile(netrcFilePath); err == nil {
+			if machine := netrc.FindMachine(machineName); machine != nil {
+				return machine
+			}
+		}
+	}
+	return nil
+}
+
+func cmdToOpenEditor(filePath string) *exec.Cmd {
+	editor := editor()
+	cmd := exec.Command(editor, filePath)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd
+}
+
+func editor() string {
+	editor := os.Getenv("EDITOR")
+	if len(editor) != 0 {
+		return editor
+	}
+	out, err := exec.Command("git", "var", "GIT_EDITOR").Output()
+	if err == nil {
+		editor = strings.TrimSpace(string(out))
+		if len(editor) != 0 {
+			return editor
+		}
+	}
+	if strings.EqualFold(runtime.GOOS, "windows") {
+		return "start"
+	} else {
+		return "vim"
+	}
+}
+
+func newTempFile(tempFileName string) (string, *os.File, error) {
+	// TODO(minwoox) change to a file in a fixed place like git
+	tempFilePath := os.TempDir() + string(filepath.Separator) + tempFileName
+	fd, err := os.Create(tempFilePath)
+	if err != nil {
+		return "", nil, errors.New("failed to create a temp file")
+	}
+	return tempFilePath, fd, nil
+}
+
+func putIntoTempFile(entry *dogma.Entry) (string, error) {
+	tempFilePath, fd, err := newTempFile(path.Base(entry.Path))
+	if err != nil {
+		return "", err
+	}
+	defer fd.Close()
+	if entry.Type == dogma.JSON {
+		b, err := marshalIndent(entry.Content)
+		if err != nil {
+			return "", err
+		}
+
+		if _, err := fd.Write(b); err != nil {
+			return "", err
+		}
+		return tempFilePath, nil
+	} else if entry.Type == dogma.Text {
+		if str, ok := entry.Content.(string); ok {
+			_, err = fd.WriteString(str)
+			if err != nil {
+				return "", err
+			}
+		} else {
+			return "", fmt.Errorf("failed to save the content: %v", entry.Content)
+		}
+	}
+	return tempFilePath, nil
+}
+
+func newUpsertChangeFromFile(fileName, repositoryPath string) (*dogma.Change, error) {
+	fileInfo, err := os.Stat(fileName)
+	if os.IsNotExist(err) {
+		return nil, fmt.Errorf("%s does not exist\n", fileName)
+	}
+	if fileInfo.IsDir() {
+		return nil, fmt.Errorf("%s is a directory\n", fileName)
+	}
+
+	change := &dogma.Change{Path: repositoryPath}
+	buf, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.HasSuffix(strings.ToLower(fileName), ".json") {
+		change.Type = dogma.UpsertJSON
+		if !json.Valid(buf) {
+			return nil, fmt.Errorf("not a valid JSON file: %s", fileName)
+		}
+		var temp interface{}
+		json.Unmarshal(buf, &temp)
+		change.Content = temp
+	} else {
+		change.Content = string(buf)
+		change.Type = dogma.UpsertText
+	}
+	return change, nil
+}
+
+func marshalIndent(data interface{}) ([]byte, error) {
+	return json.MarshalIndent(data, "", "    ")
```

Comment:
How about using 2-space indentation?

Reply:
Fixed it and corresponding examples in the document.

Reply:
Thanks!